### PR TITLE
Improving status propagation code in pipelines

### PR DESCRIPTION
**Pull request by adamfarley**
We don't seem to allow for the pipeline to have a null status, which the API tells us is possible when calling getResult().

I've patched this logic gap and added some better logging to aid future debugging.

### CODE REVIEW

**Comment by adamfarley** [comment]
Issue here: https://github.com/adoptium/ci-jenkins-pipelines/issues/750

Test run here: https://ci.adoptium.net/job/build-scripts/job/openjdk11-pipeline/2438/

**Comment by adamfarley** [comment]
The pipeline above didn't work because it was using the wrong branch, as the ci-jenkins-pipeline branch that the pipeline *itself* uses is hard-coded in the job config. I temporarily overrode the pipelines branch in the job config and launched this: https://ci.adoptium.net/job/build-scripts/job/openjdk11-pipeline/2440/

The end state of both builds in the pipeline was "unstable", and the pipeline status reflects this. Removing the WIP state of this PR, and we can worry about merging it after the release.

File: pipelines/build/common/build_base_file.groovy [hunk 1]
```diff
@@ -914,13 +914,19 @@ class Builder implements Serializable {
...
 
-                            if ((downstreamJob.getResult() != 'SUCCESS' || !copyArtifactSuccess) && propagateFailures) {
+                            if (propagateFailures) {
+                                String previousPipelineStatus = currentBuild.result
                                 context.println("Propagating downstream job result: ${downstreamJobName}, Result: "+downstreamJob.getResult()+" CopyArtifactsSuccess: "+copyArtifactSuccess)
-                                if (copyArtifactSuccess && downstreamJob.getResult() == 'UNSTABLE' && (currentBuild.result == 'SUCCESS' || currentBuild.result == 'UNSTABLE' )) {
-                                    currentBuild.result = 'UNSTABLE'
+                                if (copyArtifactSuccess) {
+                                    // currentBuild.result only allows itself to be set if the new status is worse than its current status.
+                                    // So FAILED overrides UNSTABLE, and UNSTABLE overrides PASSED.
```
Comment (andrew-m-leonard) [inline pipelines/build/common/build_base_file.groovy:922]:
```suggestion
                                    // So FAILURE overrides UNSTABLE, and UNSTABLE overrides PASSED.
```

Comment (andrew-m-leonard) [reply pipelines/build/common/build_base_file.groovy:922]:
```suggestion
                                    // So FAILURE overrides UNSTABLE, and UNSTABLE overrides SUCCESS.
```

File: pipelines/build/common/build_base_file.groovy [hunk 2]
```diff
@@ -914,13 +914,19 @@ class Builder implements Serializable {
...
 
-                            if ((downstreamJob.getResult() != 'SUCCESS' || !copyArtifactSuccess) && propagateFailures) {
+                            if (propagateFailures) {
+                                String previousPipelineStatus = currentBuild.result
                                 context.println("Propagating downstream job result: ${downstreamJobName}, Result: "+downstreamJob.getResult()+" CopyArtifactsSuccess: "+copyArtifactSuccess)
-                                if (copyArtifactSuccess && downstreamJob.getResult() == 'UNSTABLE' && (currentBuild.result == 'SUCCESS' || currentBuild.result == 'UNSTABLE' )) {
-                                    currentBuild.result = 'UNSTABLE'
+                                if (copyArtifactSuccess) {
+                                    // currentBuild.result only allows itself to be set if the new status is worse than its current status.
+                                    // So FAILURE overrides UNSTABLE, and UNSTABLE overrides PASSED.
```
Comment (adamfarley) [inline pipelines/build/common/build_base_file.groovy:922]:
```suggestion
                                    // So FAILURE overrides UNSTABLE, and UNSTABLE overrides SUCCESS.
```

**Review by andrew-m-leonard — APPROVED** [review]
lgtm

**Review by steelhead31 — APPROVED** [review]
LGTM!

**Pull request merged**